### DISCUSSION TITLE
Оптимизация таблиц циклов.

### DISCUSSION THREAD

**Discussion by MatiasShow** [totumonline/totum-ru-issues-and-discussions#614] [Как это сделать?] [answered]
В проекте в циклах таблицы циклов прилетают данные. При совершении действий insert, set или delete в таблицы циклов грузит очень долго. Как это можно оптимизировать?

**Comment by erpic-ru** [comment]
Да много как: накинуть индексы, выбрать сервер получше, оптимизировать его, асинхронировать все, что можно, сократить число выводимых записей на экран, посмотреть в профайлинге что потребляет ресурсы, оптимизировать код

**Reply by MatiasShow** [reply]
Таблицы более менее в порядке, суть в выполнении действий с внесением, изменением и удалением данных. В логах показывает больше древо пересчетов при каждом действии.

**Reply by erpic-ru** [reply]
каждое обращение к циклу вызывает пересчет расчетной таблицы. Нужно оптимизировать сам код или менять архитектуру

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Таблицы внутри цикла Рассчетные - вычисляются целиком. Если поставить код действия в поле строчной части, который будет изменять эту-же таблицу или другую таблицу в цикле (так как таблицы в цикле связаны) и изменить 10 строк, то это вызовет 10 пересчетов таблицы.

для рассчетных таблиц в таком случае используется:

1. Код действия таблицы. В коде действоя таблицы есть быстрая переменная $#changes, которая показывает какие поля и строки были изменены, добавлены, удалены и действие выполняется из кода действия таблицы.
2. Если  действие нечувствительно к строке (те важнен факт изменения общей суммы данный по колонке)то в футере колонки создается техническое поле, в которое собирается вся информация по колонке через selectList и действие вешается на изменение этого технического поля.

**Reply by ttm-alexeypolunin** [reply]
Также циклы связаны через таблицу циклов. Если работают несколько человек, ремоутов, кронов, изменяющих разные циклы одновременно, то они могут сталкиваться в таблице циклов. Что бы избежать этого в ней устанавливается тип отслеживания одновременного доступа в Без отслеживания.
